app/javascript/App/bundles/signalsPage/SignalReservationForm.tsx

```diff
@@ -0,0 +1,355 @@
+  },
+  actionsContainer: {
+    marginBottom: theme.spacing.unit * 2,
+  },
+  resetContainer: {
+    padding: theme.spacing.unit * 3,
+  },
+  fields: {
+    maxWidth: 300,
+    width: '100%',
+  },
+  textField: {},
+})
+
+interface Props {
+  classes: any
+}
+
+interface State {
+  activeStep: number
+  formData: {
+    email: string
+    telegramUsername: string
+    ethereumAddress: string
+  }
+  formErrors: {
+    email?: string[]
+    telegramUsername?: string[]
+    ethereumAddress?: string[]
+  }
+}
+
+class SignalReservationForm extends React.Component<Props, State> {
+  public STEP_COUNT = 3
+  public VALIDATION_RULES = {
+    email: 'required|email',
+    telegramUsername: [
+      'required',
+      // Telegram regex https://core.telegram.org/method/account.checkUsername
+      'regex:/^[a-zA-Z0-9_]{5,32}$/',
+    ],
+    ethereumAddress: ['required', 'regex:/^0x[a-fA-F0-9]{40}$/'],
+  }
+  public VALIDATION_ERROR_MESSAGES = {
+    required: 'Field is required',
+    email: 'Format is invalid',
+    regex: 'Format is invalid',
+  }
+
+  public state: State = {
+    activeStep: 0,
+    formData: {
+      email: '',
+      telegramUsername: '',
+      ethereumAddress: '',
+    },
+    formErrors: {},
+  }
+
+  public validateField = (fieldKey) => {
+    const rules = _.pick(this.VALIDATION_RULES, fieldKey)
+
+    // Can skip if there are no validation rules for this field
+    if (_.isEmpty(rules)) {
+      return
+    }
+
+    const validation = new Validator(
+      this.state.formData,
+      rules,
+      this.VALIDATION_ERROR_MESSAGES,
+    )
+    const validationResult = validation.passes()
+
+    this.setState((state) => ({
+      formErrors: {
+        ...state.formErrors,
+        [fieldKey]: validationResult
+          ? undefined
+          : validation.errors.get(fieldKey),
+      },
+    }))
+
+    return validationResult
+  }
+
+  public getStepFieldKeys = (stepIndex) => {
+    if (stepIndex === 0) {
+      return ['email']
+    } else if (stepIndex === 1) {
+      return ['telegramUsername']
+    } else if (stepIndex === 2) {
+      return ['ethereumAddress']
+    }
+
+    return []
+  }
+
+  public validateStep = (stepIndex) => {
+    const fieldKeys = this.getStepFieldKeys(stepIndex)
+    const haveValidationsPassed = _.every(
+      fieldKeys.map((fieldKey) => this.validateField(fieldKey)),
+    )
+
+    return haveValidationsPassed
+  }
+
+  public handleNext = () => {
+    if (!this.validateStep(this.state.activeStep)) {
+      return
+    }
+
+    const headers = {
+      'X-CSRF-Token': document
+        .querySelector('meta[name="csrf-token"]')
+        .getAttribute('content'),
+    }
+
+    axios
+      .patch(SIGNALS_RESERVATION_URL, this.state.formData, { headers })
+      .then(() => {
+        this.setState((state) => ({
+          activeStep: state.activeStep + 1,
+        }))
+      })
+      .catch((error) => {
+        // TODO: Handle error by showing some sort of message.
+        alert(error)
```

Comment:
Are we happy to leave this as an `alert`?

Reply:
Yeah, I was going to use Material UI popup but figured it's overkill for now, but we should go back and tidy it up later.

Reply:
Sure